### PR TITLE
Add max-len rule

### PR DESCRIPTION
**Pull request by NelsonBrandao**
### Rule
Add [max-len](https://eslint.org/docs/rules/max-len) rule to uphold's eslint configuration.
This will help keep track of all lengthy lines of code so they can be read more clearly. I suggest an length of `120`.

### Reason
The main reason for adding the rule is to decrease the number of code styling comments on a Pull Request. So we can focus on the important part of the review. The idea for this actually came up during the review of a Frontend PR.

*Note*: There is no `--fix` for this rule so if we decide to add it someone would need to update the code base manually. Hopefully there aren't too many of those.

What do you guys think? Feel free to add more reviewers, not really sure who I should add.

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -23,10 +23,13 @@ describe('eslint-config-uphold', () => {
   it('should generate violations for environment-specific rules', () => {
     const source = path.join(__dirname, 'fixtures', 'environment.js');
 
-    Array.from(new Set(linter.executeOnFiles([source]).results[0].messages.map(violation => violation.ruleId))).should.eql([
```
Comment (NelsonBrandao) [inline test/index.js]: A good example 😃

**Comment by Americas** [comment]
We have big screens and good resolutions, with 120 characters it feels like some screen space is just wasted :smile: 

I suggest at least 140.

**Comment by waldyrious** [comment]
I disagree with @Americas. There are multiple reasons why we may not want to make lines too long:

- Long lines are widely considered a **code smell**, since they compact a lot of complexity into a small space, making it easy to miss details and harder to grasp the intention behind the code. (It's no coincidence that one-liner C programs have typically been associated with code golf/obfuscation contests.)
- Even if the logic is clear (e.g. with chained operations), general **legibility** of long lines of text is known to be poor. Different typographical/stylistic guidelines recommend various lengths, but most are smaller than 100 characters per line (CPL), [even for electronic media](https://en.wikipedia.org/wiki/Line_length#Electronic_text).
- **Diffs** are easily made cumbersome with long lines. Side-by-side diffs suffer the most, but even inline diffs are hard to follow when lines wrap (precisely because, having different content, the soft wraps tend do happen at different locations)
- Wider screens are often taken advantage of for **multi-column development** (including having an interactive shell running in a separate column). This benefit is negated when lines get too long.

At this point, I see more substantial benefits to keeping lines to a reasonable length, than to allow them to grow just because the screen space allows it in some circumstances. So unless stronger arguments are presented for the opposite view, I strongly support this change.

**Comment by NelsonBrandao** [comment]
@waldyrious do you want to throw a number for your ideal line length? Just so we can start to see some options to decide later.

I completely agree with your points but I also think the old line length are a bit out of date for the current screen sizes. Sometimes lines are just verbose without any inherent complexity.
 
That's why we at Scytale decided on a length of 120. It felt a good compromise between screen realstate and complexity.

Btw to reduce line and function complexity there are also a couple of other rules that we can look into, like `complexity`, `max-nested-callbacks` or `max-depth`. These are way more complex and need to be used with care since they can cause a lot of programming pain if used incorrectly.

**Comment by waldyrious** [comment]
> I completely agree with your points but I also think the old line length are a bit out of date for the current screen sizes. Sometimes lines are just verbose without any inherent complexity.

Not sure which old line length you're referring to here. The venerable [72/80](https://en.wikipedia.org/wiki/Characters_per_line#In_modern_computing), perhaps?

In any case, I wouldn't suggest going that far. 120 seems fine to me.

**Comment by NelsonBrandao** [comment]
@waldyrious Yes I'm talking about the original recommend 70/80 length. Most style guides nowadays suggest around 100/120 like the Airbnb one.

**Comment by TiagoLeite07** [comment]
I agree with the 120. Not much to say, you guys said everything 👍

**Comment by rplopes** [comment]
Some of our projects are already using Prettier, which enforces a line length. Those projects are configured to use a max length of 120.

Since we've recently decided we wanted to incorporate Prettier into our projects, having an ESLint rule for the same goal sounds redundant (especially because it has no auto-fix, unlike to Prettier). So maybe we can deprecate this PR and have a new one to add Prettier with our usual setup.

WDYT?

**Comment by NelsonBrandao** [comment]
@rplopes Completely agree, when this PR was created I didn't know Prettier was already being used.

In the meantime there is already a PR open and a discussing going on about adding Prettier to `eslint-config-uphold-react`. Maybe these efforts can be combined into adding it directly on `eslint-config-uphold`.
